Telemetry blog (#510)
OctopusDeploy/blog: blog/2021-08/telemetry-crows-nest/index.md

---
title: Telemetry in Octopus Crows Nest
description: Learn how Octopus is using telemetry data to identify performance metrics for our customers
author: terence.wong@octopus.com
visibility: private # DO NOT CHANGE THIS!!!! This is not a public post!
published: 2999-01-01
metaImage: 
bannerImage: 
tags:
 - Octopus
---

> As Octopus Deploy has grown, it now covers a large software surface area. A given user may need to interface with many different touchpoints across Octopus Deploy. These may be viewing the dashboard, signing on through certificates, or viewing the projects page. To serve our customers better, we want to tell whether users accessing these touchpoints receive their content on time.

Reviewer: > it now covers a large software surface area

I'm not sure what a "software surface area" is, perhaps you mean just "surface area"? 

> interface ... touchpoints

If you audience is technical people, these are the wrong terms to use, it is management speak. `touchpoint` should be `feature` and `interface` should be `use`. I would say something more like "A typical user uses many of the different features within Octopus Deploy, for example...

> signing on through certificates

That's not something one would do. You would upload certificates.

> receive their content on time

What does on time mean? Within an hour could be on time. What we are measuring is site responsiveness. So it would be better to say the user feel the application is fast and responsive.

> Telemetry is collecting usage statistics and forwarding them to IT systems for analysis. Many software companies use it to gather data on how their customers use and experience their products. Octopus Deploy has developed a telemetry tool named Crow's Nest. We use Crow's Nest to track how quickly customers are served.

Reviewer: I would reword this to be more active. 
```suggestion
Octopus Deploy, like many software companies, collect telemetry to measure how customers experience the product. Some of the telemetry we collect is the timing of API calls and database operations, which we call `Performance Telemetry`. We surface this telemetry in our engineering dashboard (code named `Crow's Nest`).
```

> Telemetry is collecting usage statistics and forwarding them to IT systems for analysis. Many software companies use it to gather data on how their customers use and experience their products. Octopus Deploy has developed a telemetry tool named Crow's Nest. We use Crow's Nest to track how quickly customers receive web requests. Telemetry is opt-in by default unless users choose to opt-out. 

Reviewer 2: Change to emphasis what's being measured
```suggestion
Telemetry is collecting usage statistics and forwarding them to IT systems for analysis. Many software companies use it to gather data on how their customers use and experience their products. Octopus Deploy has developed a telemetry tool named Crow's Nest. We use Crow's Nest to track how quickly web requests are served. Telemetry is opt-in by default unless users choose to opt-out. 
```

> Telemetry is collecting usage statistics and forwarding them to IT systems where they for analysis. Many software companies use it to gather data on how their customers use and experience their products. Octopus Deploy has implemented a telemetry tool named Crow's Nest. Crow's Nest tracks how quickly customers receive web requests. Telemetry is opt-in by default unless users choose to opt-out. 

Reviewer 2: Can you take another look at this line, @terence-octo:

Telemetry is collecting usage statistics and forwarding them to IT systems where they for analysis.

Author: fixed

> The Crow's Nest symbolizes a tool with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 

Reviewer 2: If I'm understanding correctly, and crow's nest is a tool we developed, it might be nice to include something like:

We developed Crow's Nest to analyze....

Author: fixed

> Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 

Reviewer: This rule only applies to the headline APDEX calculated for each customer.

## Apdex

Apdex (Application Performance Index) aims to convert measurements into insights about user satisfaction. The formula is:

    Apdex = (SatisfiedCount + ToleratingCount * 0.5) / TotalCount
 
Unless in a more specific context (e.g., on individual data records or group of records), the Apdex is calculated using:

- API (Web) Requests that return a 2xx response
- Excluding certain requests that are called often and are cached (e.g., ServerStatus)
- A Satisfied Threshold of <= 50ms
- A Tolerating Threshold of > 50ms and <= 200ms

Apdex has a range of 0-100. The scale is:
- 94-100: excellent
- 85-94: good
- 70-85: fair
- below 70: poor

> - below 50: unacceptable

Reviewer: Not sure about this. The score is very context dependent. 85 overall score is probably poor.

Apdex gives a uniform scale to test the customer experience. A higher number indicates a more positive user experience.  We can vary thresholds to experiment with Apdex scores given a specific appetite for tolerance. The examples in this blog display the default threshold values. These are configurable in real-time in the application to view Apdex scores against different criteria.
 
## Visualizing Apdex and Octopus Deploy

There are several ways to visualize Apdex. The following graphs are how we use Crow's Nest to display Apdex and gain valuable insights.

### Apdex for cloud and deploy

> The blue line shows the Apdex performance of recent versions in Octopus Cloud and the deployment server. The cloud Apdex performance has been consistent at around 90 for this period. The deployment server is the internal Octopus instance not released to customers. The orange graph indicates a significant dip in Apdex from 2021.2.2048, where it recovered in 2021.2.4155. Octopus Deploy has fixed the causes of this crash and recovery. It can be helpful to look back and see how different versions affected the user experience. If there is any significant dip in performance, we can conduct a root cause analysis to identify and address causes.

Reviewer: ```suggestion
The blue line shows the Apdex performance of recent versions in Octopus Cloud and the deployment server. The cloud Apdex performance has been consistent at around 90 for this period. The deployment server is the internal Octopus instance not released to customers. The orange graph indicates a significant dip in Apdex from 2021.2.2048, where it recovered in 2021.2.4155 as we fixed the regressions. It can be helpful to look back and see how different versions affected the user experience. If there is any significant dip in performance, we can conduct a root cause analysis to identify and address causes.
```

![Apdex Cloud and Deploy](apdex-cloud-deploy.png "Apdex Cloud and Deploy")

> ### Apdex by version and license

Reviewer: I would leave out the licence type charts, they really don't give much of a signal. Presenting it like this makes it sound like this is reliable data when it's not.

Look at say the orange chart, it's all over the place for no discernible reason. If I was reading that I would wonder what was going on with our software to cause it to swing so wildly.

> This set of graphs show the Apdex score for different versions and licenses. The licenses are cloud, on-premise, trial, and overall.  The first graph below indicates that the trial subscription suffered a slight dip in Apdex performance. The performance drop was between 2020.3 and 2020.4 and between 2020.5 and 2020.6. On the x.y level, the cloud and on-premise licenses appear to be the most consistent, with the trial license suffers from more dips in performance. Overall, all licenses have Apdex levels around 85 or above, which indicates good performance.

Reviewer: > which indicates good performance
I'm just thinking about the customers who are currently frustrated with poor performance. I would work in something along the lines of "the median performance is good".

> We can filter the Apdex scores at an x.y.z and license level. The orange, blue and purple graphs below show the on-premise, cloud, and trial licenses. The blue graph confirms that the cloud license was the most consistent, maintaining around 90 over many versions. The on-premise license displayed higher peaks along with more frequent dips than the cloud license. The trial license experienced significant dips, which lowered its average Apdex score. The orange and purple graphs indicate that the on-premise and trial licenses have been less consistent than the cloud license. Trial licenses are often used for experimentation and would not be optimized for production. We are less concerned with the results of the trial license. 

Reviewer: > Trial licenses are often used for experimentation and would not be optimized for production

... and the instances they run on are likely running on less performant VMs.

> We are less concerned with the results of the trial license

I wouldn't draw that conclusion.

We can view these graphs and conduct a usability analysis between the licenses. The goal would be to bring up the usability of the on-premise licenses to match the cloud license. The latest results show the overall score above 85 and all licenses are close to one another, which is a good result.

![Apdex by Version](apdex-by-version.png "Apdex by Version")

![Apdex by Version](apdex-by-version-z-onprem.png "Apdex by Version")

![Apdex by Version](apdex-by-version-z-cloud.png "Apdex by Version")

![Apdex by Version](apdex-by-version-z-trial.png "Apdex by Version")

<!--### Apdex customer view

![Apdex Customer View](apdex-customer.png "Apdex Customer View")-->

### Apdex overall score

> Each customer has an overall Apdex score. Apdex scores below a certain level prompt engineers to fix any faults causing the poor score.

Reviewer: > Apdex scores below a certain level prompt engineers to fix any faults causing the poor score.

That is not true

![Apdex Score](apdex-score.png "Apdex Score")

### Apdex routes

> Every web request has an endpoint. The endpoint is the subject of the request. The path to fulfilling these requests are routes. The performance of individual routes can are in the customer's view. The view shows metrics such as the mean, median and highest value of a request. We can use these metrics to identify the worst-performing routes in a later version.

Reviewer: I would drop the concept of `endpoint` here. `route` is the best term to use.

"We time web request and group the results on the route (path). From these timing we show the mean, median and 95th percentile time".

It would be interesting to talk about how we bucket the requests so that we don't send up huge amounts of data.

![Apdex Routes](apdex-route.png "Apdex Routes")

#### Apdex route difference

We can use Crow's Nest to view the differences in routes across different versions. The project routes improved between 2020.6 and 2021.1, as indicated by the green difference indicators for the Apdex score.

![Apdex Routes Difference](apdex-route-diff.png "Apdex Routes Difference")

#### Apdex route view

> Historical performance is visible on every route. This is useful when assessing whether updated versions of Octopus Deploy have improved or worsened the route. The performance of the certificates route above degraded from 2020.2 to 2020.6. The Apdex score decreased from 62 to 16. In 2021.1, this route improved, leading to an Apdex score of 77.

Reviewer: ```suggestion
Historical performance is visible on every route. This allows us to see how the performance of a route has changed in each release. For example, the performance of the certificates route above degraded from 2020.2 to 2020.6. The Apdex score decreased from 62 to 16. In 2021.1, this cause of this slowness was resolved, leading to an Apdex score of 77.
```

![Apdex Route View](apdex-route-view.png "Apdex Route View")

#### Future improvements

> We have found this tool to be a source of information that can highlight routes for improvement. This lets us hone in on areas where we can improve the apdex scores, improving the quality of the Octopus Deploy product. 

Reviewer: We have found that this tool helps us plan on what to work on next. It highlights the areas that need the most attention and should be addressed first. Also it gives us another signal when monitoring our overall application health.

Telemetry is collecting usage statistics and forwarding them to IT systems where they for analysis. Many software companies use it to gather data on how their customers use and experience their products. Octopus Deploy has implemented a telemetry tool named Crow's Nest. Crow's Nest tracks how quickly customers receive web requests. Telemetry is opt-in by default unless users choose to opt-out. 

The Crow's Nest symbolizes a tool with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 

Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 

> ### Apdex

Reviewer 2: ```suggestion
## Apdex
```

author: terence.wong@octopus.com
visibility: private # DO NOT CHANGE THIS!!!! This is not a public post!
published: 2999-01-01
metaImage: 
bannerImage: 
tags:
 - Octopus
---

As Octopus Deploy has grown, it now covers a large surface area. A given user may need to use different features across Octopus Deploy. These may be viewing the dashboard, creating or deploying releases, authoring projects or running runbooks against their infrastructure. To serve our customers better, we want to tell whether users receive a fast and responsive experience.

Octopus Deploy, like many software companies, collect telemetry to measure how customers experience the product. Some of the telemetry we collect is the timing of API calls and database operations, which we call 'Performance Telemetry'. We surface this telemetry in our engineering dashboard (code named Crow's Nest).

Crow's Nest provides us with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 

## Apdex

> Apdex (Application Performance Index) aims to convert measurements into insights about user satisfaction. The formula is:

Reviewer 2: ```suggestion
We calculate an Apdex (Application Performance Index) score, which aims to convert measurements into insights about user satisfaction. The formula is:
```

    Apdex = (SatisfiedCount + ToleratingCount * 0.5) / TotalCount
 

> Unless in a more specific context (e.g., on individual data records or group of records), the Apdex is calculated using:

Reviewer 2: ```suggestion
The Apdex is calculated using:
```

---

> title: Telemetry in Octopus Crows Nest

Reviewer 2: ```suggestion
title: How we use Telemetry to improve Octopus Deploy
```

> description: Learn how Octopus is using telemetry data to identify performance metrics for our customers

Reviewer 2: ```suggestion
description: Learn how Octopus is using telemetry data to identify performance metrics and improve Octopus Deploy for our customers
```

author: terence.wong@octopus.com
visibility: private # DO NOT CHANGE THIS!!!! This is not a public post!
published: 2999-01-01
metaImage: 
bannerImage: 
tags:
 - Octopus
---

> As Octopus Deploy has grown, it now covers a large surface area. A given user may need to use different features across Octopus Deploy. These may be viewing the dashboard, creating or deploying releases, authoring projects or running runbooks against their infrastructure. To serve our customers better, we want to tell whether users receive a fast and responsive experience.

Reviewer 2: ```suggestion
Over the years, Octopus Deploy has grown to include many new features and areas within the app.  Users may need to use different features across Octopus Deploy, including viewing the dashboard, creating or deploying releases, authoring projects or running runbooks against their infrastructure. To serve our customers better, we want to know how   fast and responsive the experience is.
```

> Octopus Deploy, like many software companies, collect telemetry to measure how customers experience the product. Some of the telemetry we collect is the timing of API calls and database operations, which we call 'Performance Telemetry'. We surface this telemetry in our engineering dashboard (code named Crow's Nest).

Reviewer 2: ```suggestion
Octopus Deploy, like many software companies, collect telemetry to measure how customers use the product. Some of the telemetry we collect is the timing of API calls and database operations, which we call 'Performance Telemetry'. We surface this telemetry in our engineering dashboard (code named Crow's Nest).
```

> Crow's Nest provides us with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 

Reviewer 2: ```suggestion
Crow's Nest provides us with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint, and Crows Nest measure how long the request takes. 

We also track these requests over several versions to ensure Octopus Deploy's responsiveness continues to improve. 
```

> The Crow's Nest symbolizes a tool with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 

Reviewer 2: ```suggestion
Crow's Nest provides us with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint. These requests get tracked over several versions. 
```

> Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 

Reviewer 2: ```suggestion
```

## Apdex

Apdex (Application Performance Index) aims to convert measurements into insights about user satisfaction. The formula is:

    Apdex = (SatisfiedCount + ToleratingCount * 0.5) / TotalCount
 
Unless in a more specific context (e.g., on individual data records or group of records), the Apdex is calculated using:

- API (Web) Requests that return a 2xx response
- Excluding certain requests that are called often and are cached (e.g., ServerStatus)
- A Satisfied Threshold of <= 50ms
- A Tolerating Threshold of > 50ms and <= 200ms

Apdex gives a uniform scale to test the customer experience. A higher number indicates a more positive user experience.  We can vary thresholds to experiment with Apdex scores given a specific appetite for tolerance. The examples in this blog display the default threshold values. These are configurable in real-time in the application to view Apdex scores against different criteria.

>  

Reviewer 2: ```suggestion
 Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 
```

Octopus Deploy, like many software companies, collect telemetry to measure how customers experience the product. Some of the telemetry we collect is the timing of API calls and database operations, which we call 'Performance Telemetry'. We surface this telemetry in our engineering dashboard (code named Crow's Nest).

Crow's Nest provides us with a high-level overview of how users are experiencing the product. A web request triggers when a customer wants to visit a page or poll an endpoint, and Crows Nest measure how long the request takes. 

We also track these requests over several versions to ensure Octopus Deploy's responsiveness continues to improve. 

## Apdex

Apdex (Application Performance Index) aims to convert measurements into insights about user satisfaction. The formula is:

    Apdex = (SatisfiedCount + ToleratingCount * 0.5) / TotalCount
 
Unless in a more specific context (e.g., on individual data records or group of records), the Apdex is calculated using:

- API (Web) Requests that return a 2xx response

> - Excluding certain requests that are called often and are cached (e.g., ServerStatus)

Reviewer 2: ```suggestion
- Excludes certain requests that are called often and are cached
```

> - A Satisfied Threshold of <= 50ms

Reviewer 2: ```suggestion
- Requests that are returned in less than or equal to 50ms are considered to be with the satisfied threshold
```

> - A Tolerating Threshold of > 50ms and <= 200ms

Reviewer 2: ```suggestion
- Requests that are greater than 50ms and less than 200ms are considered to be within a tolerating threshold
```

Apdex gives a uniform scale to test the customer experience. A higher number indicates a more positive user experience.  We can vary thresholds to experiment with Apdex scores given a specific appetite for tolerance. The examples in this blog display the default threshold values. These are configurable in real-time in the application to view Apdex scores against different criteria.

>  Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 

Reviewer 2: ```suggestion

Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 
```

We also track these requests over several versions to ensure Octopus Deploy's responsiveness continues to improve. 

## Apdex

Apdex (Application Performance Index) aims to convert measurements into insights about user satisfaction. The formula is:

    Apdex = (SatisfiedCount + ToleratingCount * 0.5) / TotalCount
 
Unless in a more specific context (e.g., on individual data records or group of records), the Apdex is calculated using:

- API (Web) Requests that return a 2xx response
- Excludes certain requests that are called often and are cached
- A Satisfied Threshold of <= 50ms
- A Tolerating Threshold of > 50ms and <= 200ms

Apdex gives a uniform scale to test the customer experience. A higher number indicates a more positive user experience.  We can vary thresholds to experiment with Apdex scores given a specific appetite for tolerance. The examples in this blog display the default threshold values. These are configurable in real-time in the application to view Apdex scores against different criteria.

> Only web requests with a 2xx status are in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 

Reviewer 2: ```suggestion
Only web requests with a 2xx status are included in the calculations. A 2xx status indicates a successful web request. Versions with less than 50 instances sending telemetry on any given data get filtered out to remove outliers. The response times of these web requests estimate how satisfied a customer is with their service. 
```

![Apdex Cloud and Deploy](apdex-cloud-deploy.png "Apdex Cloud and Deploy")

### Apdex by version and license

We can compare the Apdex score for different versions and licenses. The licenses are cloud, on-premise, trial, and overall. There is a lot of variability in these numbers as they are representative of all customers. 

![Apdex by Version](apdex-by-version.png "Apdex by Version")

<!--### Apdex customer view

![Apdex Customer View](apdex-customer.png "Apdex Customer View")-->

### Apdex overall score

Each customer has an overall Apdex score. This score gives an indication of the responsiveness of the customer experience.

![Apdex Score](apdex-score.png "Apdex Score")

### Apdex routes

> The performance of individual routes can are in the customer's view. We time web request and group the results on the route (path). From these timings we show the mean, median and 95th percentile time. By bucketing the requests by route, we avoid sending up huge amounts of data. We can use these metrics to identify the worst-performing routes in a later version. 

Reviewer 2: ```suggestion
The performance of individual routes are in the customer view. We time web request and group the results on the route. From these timings we show the mean, median and 95th percentile time. By bucketing the requests by route, we avoid transmitting huge amounts of data. We can use these metrics to identify the worst-performing routes in a later version. 
```

![Apdex Routes](apdex-route.png "Apdex Routes")

#### Apdex route difference

We can use Crow's Nest to view the differences in routes across different versions. The project routes improved between 2020.6 and 2021.1, as indicated by the green difference indicators for the Apdex score.

![Apdex Routes Difference](apdex-route-diff.png "Apdex Routes Difference")

#### Apdex route view

> Historical performance is visible on every route. This allows us to see how the performance of a route has changed in each release. For example, the performance of the certificates route above degraded from 2020.2 to 2020.6. The Apdex score decreased from 62 to 16. In 2021.1, this cause of this slowness was resolved, leading to an Apdex score of 77.

Reviewer 2: ```suggestion
Historical performance is visible on every route. This allows us to see how the performance of a route has changed in each release. For example, the performance of the certificates route above degraded from 2020.2 to 2020.6. The Apdex score decreased from 62 to 16. In 2021.1, we identified the cause of this decrease in Apdex score and resolved the issue, leading to an Apdex score of 77.
```

![Apdex Route View](apdex-route-view.png "Apdex Route View")

#### Future improvements

> We have found that this tool helps us plan on what to work on next. It highlights the areas that need the most attention and should be addressed first. Also it gives us another signal when monitoring our overall application health.

Reviewer 2: ```suggestion
We have found that Crow's Nest helps us plan what to work on next. It highlights the areas that need the most attention and should be addressed first. It also gives us another signal when monitoring our overall application health.
```

<!--![Apdex Routes Difference](apdex-route-dashboard.png "Apdex Routes Difference")-->

> Telemetry is a powerful tool that can enable businesses to gain a complete picture of user experiences. Telemetry and Apdex give full visibility of each user and the performance of each route. We can compare performance across different versions and licenses. Comparing versions quantifies the effect of each update. Poor-performing routes are visible on a per-user basis or across the Octopus Deploy platform. Our Crow's Nest tool helps to improve the customer experience, leading to customer satisfaction.

Reviewer 2: ```suggestion
Telemetry is a powerful tool that can enable businesses to gain a complete picture of user experiences. Telemetry and Apdex give full visibility of each user and the performance of each route. We can compare performance across different versions and licenses. Comparing versions quantifies the effect of each update. Poor-performing routes are visible on a per-user basis or across the Octopus Deploy platform. Our Crow's Nest tool helps to improve the customer experience, leading to a more responsive experience and happier deployments.
```
